src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs

```diff
@@ -18539,5 +18537,171 @@ public struct A
                 //     public A(int x, A a = M($"{1}")) { }
                 Diagnostic(ErrorCode.ERR_BadArgRef, @"$""{1}""").WithArguments("1", "ref").WithLocation(8, 29));
         }
+
+        [Theory]
+        [CombinatorialData]
+        public void InterpolatedStringHandlerArgumentAttribute_ThisParameter_ReferenceTypeReferenceReceiver(
+            [CombinatorialValues(@"$""literal""", @"$""literal"" + $""""")] string expression)
+        {
+            var code = $$"""
+using System;
+using System.Runtime.CompilerServices;
+C c = new C(5);
+ref C c2 = ref c;
+C otherC = null;
+c2.M({{expression}}, c2 = ref otherC);
```

Comment:
>ref otherC [](http://example.com/codeflow?start=26&length=10)

I do not think this scenario is interesting. I.e. redirecting `c2` to a different location is not interesting because the old implementation was capturing the reference and changing that reference couldn't have any effect afterwards. What would be interesting is a change to a value that the reference points to, ```c2 = otherC```. After making this change to the test scenarios in this unit-test and the unit-test below, we will be able to observe that the implementation change is not merely an optimization, but actually a fix for a real problem. Scenario from `InterpolatedStringHandlerArgumentAttribute_ThisParameter_ReferenceTypeReferenceReceiver_ReverseOrder` is going to crash at runtime without the implementation change and with the implementation change it is going to exhibit expected behavior. Therefore, I suggest opening an issue with a "broken" repro, adjusting description of this PR to say that we are attempting to fix that issue, and adjusting the added unit-tests accordingly.

Reply:
I would even suggest to simplify the scenario by using ```c2 = default``` and getting rid of `otherC` local

Reply:
Also, please add tests for generic scenarios like:
```
            var code = $$"""
using System;
using System.Runtime.CompilerServices;

class Program
{
    static void Main()
    {
        C c = new C(5);
        Test(ref c);
    }

    static void Test<T>(ref T c2) where T : I1
    {
        c2.M(c2 = default, {{expression}});
    }
}

public interface I1
{
    int Prop { get; }
    void M(I1 cNull, [InterpolatedStringHandlerArgumentAttribute("")] CustomHandler c);
}

public class C : I1
{
    public int Prop { get; }
    public C(int i) => Prop = i;
    public void M(I1 cNull, [InterpolatedStringHandlerArgumentAttribute("")] CustomHandler c)
    {
        Console.WriteLine(cNull is null);
        Console.WriteLine(c.ToString());
    }
}

public partial struct CustomHandler
{
    public CustomHandler(int literalLength, int formattedCount, I1 c) : this(literalLength, formattedCount)
    {
        _builder.AppendLine("c.Prop:" + c.Prop.ToString());
    }
}
""";
```

and

```
            var code = $$"""
using System;
using System.Runtime.CompilerServices;

class Program
{
    static void Main()
    {
        C c = new C(5);
        Test(ref c);
    }

    static void Test<T>(ref T c2) where T : class, I1
    {
        c2.M(c2 = default, {{expression}});
    }
}

public interface I1
{
    int Prop { get; }
    void M(I1 cNull, [InterpolatedStringHandlerArgumentAttribute("")] CustomHandler c);
}

public class C : I1
{
    public int Prop { get; }
    public C(int i) => Prop = i;
    public void M(I1 cNull, [InterpolatedStringHandlerArgumentAttribute("")] CustomHandler c)
    {
        Console.WriteLine(cNull is null);
        Console.WriteLine(c.ToString());
    }
}

public partial struct CustomHandler
{
    public CustomHandler(int literalLength, int formattedCount, I1 c) : this(literalLength, formattedCount)
    {
        _builder.AppendLine("c.Prop:" + c.Prop.ToString());
    }
}
""";
```